### PR TITLE
Image handler API merge

### PR DESCRIPTION
**Pull request by tavisca-rpandey**

### CODE REVIEW

File: src/OnlineRetailPortal.Contracts/Contracts/DeleteImageRequest.cs [hunk 1]
```diff
@@ -0,0 +1,11 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace OnlineRetailPortal.Contracts.Contracts
+{
+    public class DeleteImageRequest
+    {
+        public string ImageID { get; set; }
```
Comment (tavisca-swapnil) [inline src/OnlineRetailPortal.Contracts/Contracts/DeleteImageRequest.cs]: Rename to ImageId

File: src/OnlineRetailPortal.Web/Controllers/ImagesController.cs [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+            _imageHandler = imageHandler;
+        }
+
+        // POST: api/Images
+        [HttpPost]
+        public async Task<IActionResult> UploadImage()
+        {
+            ImageRequestValidator.ValidateImagePostRequest(Request);
+
+            IFormFile file = Request.Form.Files[0];
```
Comment (tavisca-swapnil) [inline src/OnlineRetailPortal.Web/Controllers/ImagesController.cs]: No logic in controller.

File: src/OnlineRetailPortal.Web/Controllers/ImagesController.cs [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+            UploadImageResponse response = await _imageHandler.UploadImage(file.ToUploadImageServiceContract());
+            return response.ToUser();
+            
+        }
+
+
+        // DELETE: api/ApiWithActions/5
+        [HttpDelete("{id}")]
+        public IActionResult DeleteImage(string id)
+        {
```
Comment (tavisca-swapnil) [inline src/OnlineRetailPortal.Web/Controllers/ImagesController.cs:40]: Let's validate 'id' here.

File: src/OnlineRetailPortal.Web/Validations/UploadImageRequestValidator.cs [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+            .Cascade(CascadeMode.StopOnFirstFailure)
+            .NotNull()
+            .WithErrorCode(StatusCodes.Status400BadRequest.ToString())
+            .WithMessage(Error.MissingImage());
+
+
+
+            RuleFor(req => req.Form.Files[0])
+            .Cascade(CascadeMode.StopOnFirstFailure)
+            .Must(file => IsImageFile(file))
```
Comment (tavisca-swapnil) [inline src/OnlineRetailPortal.Web/Validations/UploadImageRequestValidator.cs]: Rename IsImageFile to IsSupportedImageFile

File: src/OnlineRetailPortal.Contracts/Contracts/IImageHandler.cs [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+namespace OnlineRetailPortal.Contracts.Contracts
+{
+    public interface IImageHandler
+    {
+        /// <summary>
+        /// Takes an UploadImageRequest type object and tries to save the image in it
+        /// </summary>
+        /// <param name="request"></param>
+        /// <returns></returns>
+        Task<UploadImageResponse> UploadImage(UploadImageRequest request);
```
Comment (Tavisca-smohite) [inline src/OnlineRetailPortal.Contracts/Contracts/IImageHandler.cs]: the methods which are returning tasks (async methods) should be suffixed with `async`.

File: src/OnlineRetailPortal.Contracts/Contracts/IImageHandler.cs [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+        /// <summary>
+        /// Takes a DeleteImageRequest object containing the name of the file to delete from the temporary folder
+        /// </summary>
+        /// <param name="imageUrl"></param>
```
Comment (Tavisca-smohite) [inline src/OnlineRetailPortal.Contracts/Contracts/IImageHandler.cs]: this param is not part of your contract right ?

File: src/OnlineRetailPortal.Contracts/Contracts/UploadImageRequest.cs [hunk 7]
```diff
@@ -0,0 +1,14 @@
+﻿using FluentValidation;
+using Microsoft.AspNetCore.Http;
+using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace OnlineRetailPortal.Contracts.Contracts
```
Comment (Tavisca-smohite) [inline src/OnlineRetailPortal.Contracts/Contracts/UploadImageRequest.cs]: resolve the namespace to ` OnlineRetailPortal.Contracts`

File: src/OnlineRetailPortal.Web/Validations/UploadImageRequestValidator.cs [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+            RuleFor(req => req)
+           .Cascade(CascadeMode.StopOnFirstFailure)
+           .NotNull()
+           .WithErrorCode(StatusCodes.Status400BadRequest.ToString())
+           .WithMessage(Error.MissingImage());
+
+
+
+            RuleFor(req => req.Form)
+            .Cascade(CascadeMode.StopOnFirstFailure)
```
Comment (Tavisca-smohite) [inline src/OnlineRetailPortal.Web/Validations/UploadImageRequestValidator.cs]: why CascadeMode if the rule has only one validation ?

**Review by Tavisca-smohite — CHANGES_REQUESTED** [review]

File: src/OnlineRetailPortal.Contracts/Contracts/UploadImageRequest.cs (hunk 7 above)
Comment (tavisca-rpandey) [reply src/OnlineRetailPortal.Contracts/Contracts/UploadImageRequest.cs]: Okay, Done

File: src/OnlineRetailPortal.Contracts/Contracts/IImageHandler.cs (hunk 5 above)
Comment (tavisca-rpandey) [reply src/OnlineRetailPortal.Contracts/Contracts/IImageHandler.cs]: Okay, Done

File: src/OnlineRetailPortal.Web/Validations/UploadImageRequestValidator.cs (hunk 8 above)
Comment (tavisca-rpandey) [reply src/OnlineRetailPortal.Web/Validations/UploadImageRequestValidator.cs]: Yes, Removed

File: src/OnlineRetailPortal.Contracts/Contracts/IImageHandler.cs (hunk 6 above)
Comment (tavisca-rpandey) [reply src/OnlineRetailPortal.Contracts/Contracts/IImageHandler.cs]: Done

File: src/OnlineRetailPortal.Services/Services/ImageService.cs [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+        }
+
+        public List<string> MoveToStorage(List<string> images)
+        {
+            List<string> storedImages = new List<string>();
+            string tempPath = Path.Combine(Directory.GetCurrentDirectory(), _env.WebRootPath, _tempImageFolder);
+            string storagePath = Path.Combine(Directory.GetCurrentDirectory(), _env.WebRootPath, _storageFolder);
+            try 
+            {
+                foreach (string image in images)
```
Comment (tavisca-swapnil) [inline src/OnlineRetailPortal.Services/Services/ImageService.cs]: Add null check and return null if images==null

File: src/OnlineRetailPortal.Services/Services/ImageService.cs [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+            {
+                //Logger.logInformation("Invalid Delete request: {@ex} ", ex)
+                throw new BaseException(StatusCodes.Status500InternalServerError, "Internal Server Error",null, System.Net.HttpStatusCode.InternalServerError);
+            }
+
+        }
+
+        public List<string> MoveToStorage(List<string> images)
+        {
+            List<string> storedImages = new List<string>();
```
Comment (tavisca-swapnil) [inline src/OnlineRetailPortal.Services/Services/ImageService.cs]: Rename storedImages

File: src/OnlineRetailPortal.Services/Services/ImageService.cs [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+        /// Takes an imageID inside a DeleteImageRequest Object and deletes the image from the temporary Storage
+        /// </summary>
+        /// <param name="request"></param>
+        public void DeleteTempImage(DeleteImageRequest request)
+        {
+            string imageId = request.ImageId;
+            string imageFolder = GetImageFolder(imageId); 
+            try
+            {
+                string path = Path.Combine(Directory.GetCurrentDirectory(), _env.WebRootPath, _tempImageFolder, imageId);
```
Comment (tavisca-swapnil) [inline src/OnlineRetailPortal.Services/Services/ImageService.cs:49]: Replace _tempImageFolder with imageFolder

**Review by tavisca-swapnil — CHANGES_REQUESTED** [review]

**Review by tavisca-swapnil — APPROVED** [review]

File: src/OnlineRetailPortal.Contracts/Contracts/DeleteImageRequest.cs (hunk 1 above)
Comment (tavisca-rpandey) [reply src/OnlineRetailPortal.Contracts/Contracts/DeleteImageRequest.cs]: Okay done.

File: src/OnlineRetailPortal.Web/Controllers/ImagesController.cs (hunk 2 above)
Comment (tavisca-rpandey) [reply src/OnlineRetailPortal.Web/Controllers/ImagesController.cs]: Okay done.

File: src/OnlineRetailPortal.Web/Controllers/ImagesController.cs (hunk 3 above)
Comment (tavisca-rpandey) [reply src/OnlineRetailPortal.Web/Controllers/ImagesController.cs:40]: Okay done.

File: src/OnlineRetailPortal.Web/Validations/UploadImageRequestValidator.cs (hunk 4 above)
Comment (tavisca-rpandey) [reply src/OnlineRetailPortal.Web/Validations/UploadImageRequestValidator.cs]: Okay done.

File: src/OnlineRetailPortal.Services/Services/ImageService.cs (hunk 9 above)
Comment (tavisca-rpandey) [reply src/OnlineRetailPortal.Services/Services/ImageService.cs]: Okay done.

File: src/OnlineRetailPortal.Services/Services/ImageService.cs (hunk 10 above)
Comment (tavisca-rpandey) [reply src/OnlineRetailPortal.Services/Services/ImageService.cs]: Okay done.

File: src/OnlineRetailPortal.Services/Services/ImageService.cs (hunk 11 above)
Comment (tavisca-rpandey) [reply src/OnlineRetailPortal.Services/Services/ImageService.cs:49]: Okay done.

File: src/OnlineRetailPortal.Web/Controllers/ImagesController.cs [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+
+
+        // DELETE: api/ApiWithActions/5
+        [HttpDelete("images/{id}")]
+        public IActionResult DeleteImage(string id)
+        {
+            DeleteImageRequestValidator requestValidator = new DeleteImageRequestValidator();
+            requestValidator.ValidateDeleteRequest(id);
+
+            _imageService.DeleteTempImage(id.ToDeleteImageContract());
```
Comment (Tavisca-smohite) [inline src/OnlineRetailPortal.Web/Controllers/ImagesController.cs]: why image service has the method named `DeleteTempImage`

Comment (tavisca-rpandey) [reply src/OnlineRetailPortal.Web/Controllers/ImagesController.cs]: Changed the name, had refactored the code but forgot to rename.

Comment (tavisca-rpandey) [reply src/OnlineRetailPortal.Web/Controllers/ImagesController.cs]: done

**Review by Tavisca-smohite — APPROVED** [review]

**Review by achirametavisca — APPROVED** [review]

**Pull request merged**